tests/testthat/test-generate_data.R

```diff
@@ -381,3 +378,282 @@ test_that("genDummy works.", {
+    
+    expect_equal(as.numeric(gmnw), as.numeric(gmw))
+  }
+  
+  for (i in 1:5) {
+    check_equal(gm1, gm1_w)
+    check_equal(gm2, gm2_w)
+  }
+  
+  set.seed(oldSeed)
+  
+})
+
+#genMultiFac ----
+test_that("genMultiFac throws errors.", {
+  oldSeed <- .Random.seed
+  set.seed(98765)
+  
+  # check nFactors are integers
+  expect_error(genMultiFac(1.4, each = 4), class = "simstudy::wrongType")
+  
+  # check length nFactors greater than 2
+  expect_error(genMultiFac(1, each = 4), class = "simstudy::greaterThan")
+  
+  # check number of levels matches factors
+  expect_error(genMultiFac(3, levels = c(2, 3)), class = "simstudy::notEqual")
+  
+  # check coding == 'effect' or 'dummy'
+  expect_error(genMultiFac(2, each = 3, coding = "trtAssign"), class = "simstudy::codingVal")
+  
+  set.seed(oldSeed)
+})
+
+test_that("genMultiFac works.", {
+  oldSeed <- .Random.seed
+  set.seed(87654)
+  
+  # coding == dummy, levels == 2
+  nFac <- sample(2:5, size = 1)
+  nEach <- sample(2:5, size = 1)
+  g1 <- genMultiFac(nFac, each = nEach)
+  g1
+  
+  # checks each column sums to correct amount
+  for(i in sample(2:(nFac + 1))) {
+    expect_equal(sum(g1[, i, with = FALSE]), nrow(g1) / 2)
+  }
+  
+  # checks all values are 0s or 1s
+  expect_true(all(g1[, 2:(nFac + 1)] == 0 | g1[, 2:(nFac + 1)] == 1))
+  
+  # checks all rows are unique
+  rowStrings <- unlist(lapply(split(g1[,-1], seq(nrow(g1))), function(x) paste0(x, collapse="")))
+  expect_true(length(rowStrings) == (length(unique(rowStrings)) * nEach))
+  
+  # checks there is right number of rows
+  expect_true(length(rowStrings) == nEach * 2^nFac)
+  
+  
+  ## coding == effect, levels == 2
+  nFac <- sample(2:5, size = 1)
+  nEach <- sample(2:5, size = 1)
+  g2 <- genMultiFac(nFac, each = nEach, coding = "effect")
+  
+  # checks all columns sum to 0
+  for(i in sample(2:(nFac + 1))) {
+    expect_equal(sum(g2[, i, with = FALSE]), 0)
+  }
+  
+  # checks all values are 1s or -1s
+  expect_true(all(g2[, 2:(nFac + 1)] == 1 | g2[, 2:(nFac + 1)] == -1))
+  
+  # checks all rows are unique
+  rowStrings <- unlist(lapply(split(g2[,-1], seq(nrow(g2))), function(x) paste0(x, collapse="")))
+  expect_true(length(rowStrings) == (length(unique(rowStrings)) * nEach))
+  
+  # checks there is right number of rows
+  expect_true(length(rowStrings) == nEach * 2^nFac)
+  
+  
+  ## levels == other, len(levels) == 1
+  nFac <- sample(2:5, size = 1)
+  nEach <- sample(2:5, size = 1)
+  nLev <- sample(2:5, size = 1)
+  g3 <- genMultiFac(nFac, each = nEach, levels = nLev)
+  
+  # checks all rows are unique
+  rowStrings <- unlist(lapply(split(g3[,-1], seq(nrow(g3))), function(x) paste0(x, collapse="")))
+  expect_true(length(rowStrings) == (length(unique(rowStrings)) * nEach))
+  
+  # checks there is right number of rows
+  expect_true(length(rowStrings) == nEach * nLev^nFac)
+  
+  # checks all values are in correct range
+  expect_true(all(g3[, i, with = FALSE] <= nLev[i - 1]))
```

Comment:
Since all columns should have the same parameters in terms of maximum and minimum values, I changed this to just observe the whole table (excluding the IDs)